server/storage/mvcc/key_index.go

```diff
@@ -262,8 +257,15 @@ func (ki *keyIndex) doCompact(atRev int64, available map[Revision]struct{}) (gen
 	genIdx, g := 0, &ki.generations[0]
 	// find first generation includes atRev or created after atRev
 	for genIdx < len(ki.generations)-1 {
-		if tomb := g.revs[len(g.revs)-1].Main; tomb > atRev {
-			break
+		// if len(g.revs) == 1, then it means that the generation has
+		// only one tombstone revision, which was kept in last compact
+		// operation. We need to clean it up in this round of compact
+		// operation. Refer to discussion in
+		// https://github.com/etcd-io/etcd/issues/18089.
+		if len(g.revs) > 1 {
```

Comment:
I think we don't need this if here. Just let it check the following condition. It's more easy to understand and reason

Reply:
It seems right.  If `len(g.revs)==1` is `true`, it means that lastCompactionRev must be equal to the tombstone rev. In this round of compaction, the tombstone revision must be < compaction rev.

So we can revert this change.